src/rand.rs

```diff
@@ -83,21 +84,64 @@ pub(crate) mod sealed {
         fn as_mut_bytes(&mut self) -> &mut [u8]; // `AsMut<[u8]>::as_mut`
     }
 
-    macro_rules! impl_random_arrays {
-        [ $($len:expr)+ ] => {
-            $(
-                impl RandomlyConstructable for [u8; $len] {
-                    #[inline]
-                    fn zero() -> Self { [0; $len] }
+    // Safety: in order for a type to be POD, it must satisfy the following criteria:
+    // - The type must be inhabited (ie not the never type)
+    // - All bit patterns must be valid
+    // - The type must have no padding
+    // - Any fields within the type must also be Pod
+    pub unsafe trait Pod: Sized {
+        fn zero() -> Self;
+    }
 
+    macro_rules! impl_pod {
+        ($($type:ty),*) => {
+            $(
+                unsafe impl Pod for $type {
                     #[inline]
-                    fn as_mut_bytes(&mut self) -> &mut [u8] { &mut self[..] }
+                    fn zero() -> Self { 0 }
                 }
-            )+
-        }
+            )*
+        };
     }
 
-    impl_random_arrays![4 8 16 32 48 64 128 256];
+    impl_pod!(i8, u8, i16, u16, i32, u32, i64, u64, i128, u128, isize, usize);
+
+    impl<T: Pod + Copy, const N: usize> RandomlyConstructable for [T; N] {
+        #[inline]
+        fn zero() -> Self {
+            // Note that the Copy bound on T is required in order to initialize this array
+            // without unsafe code.
+            [T::zero(); N]
+        }
+
+        #[inline]
+        fn as_mut_bytes(&mut self) -> &mut [u8] {
+            // Since all values in this expression are constant, the compiler should optimize this
+            // assertion out.
+            assert!(
+                N == 0 || (isize::MAX as usize) / N >= mem::size_of::<T>(),
+                "Array too large to be converted to slice of u8"
+            );
+
+            let data = self.as_mut_ptr();
+            let len = N * mem::size_of::<T>();
+
+            // Safety:
+            // - `data` is valid (aligned, non-null, non-dangling) for reads and writes of size
+            //   `len` bytes
+            //   - The entire memory range of this slice is in a single allocated object because
+            //     it came from another mutable slice
+            //   - `data` is not dangling and properly aligned for all types because, again, it was
+            //     derived from another mutable slice
+            // - Because T is POD, and because `len` is calculated based on the size of T,
+            //   `data` points to `len` initialized u8s
+            // - We co-opt the lifetime of `self`, so no borrow checker rules are broken
```

Comment:
I'm not sure about this:
>             // - We co-opt the lifetime of `self`, so no borrow checker rules are broken

It's hard for me to remember the exact circumstances, but I recall trying to do something similar and finding that I was running into a conflict with the mutable noalias optimization or similar.

This, in conjunction with wanting to avoid the need to initially zero all the bytes to start with, made me want to rewrite this whole module in terms of `MaybeUninit` and pointers.

WDYT?

Reply:
So I had not heard of the mutable noalias optimization before, and quick google searches of it suggest that it's actually [buggy in the LLVM](https://stackoverflow.com/a/57259339/5619133). In a world where there is no mis-compilation, the function signature should essentially be equivalent to `fn as_mut_bytes<'a>(&'a mut self) -> &'a mut [u8]`, and since `'a` will be inferred by the usage of the return value, it will extend the borrow of `self` accordingly. I can ask some other rustaceans I know about it to double check though.

When reading through the module I also thought about what it would look like to switch it over to using raw pointers. I think that's a viable option (and skips the pain of creating an initially zeroed-out array), but I didn't want to get waist deep with changing `SecureRandom` on my first contribution.